lms/djangoapps/course_blocks/transformers/library_content.py

```diff
@@ -0,0 +1,176 @@
+                filter_func=lambda block_key: block_key.block_type == 'library_content',
+                yield_descendants_of_unyielded=True,
+        ):
+            library_children = block_structure.get_children(block_key)
+            if library_children:
+                all_library_children.update(library_children)
+                selected = []
+                mode = block_structure.get_xblock_field(block_key, 'mode')
+                max_count = block_structure.get_xblock_field(block_key, 'max_count')
+
+                # Retrieve "selected" json from LMS MySQL database.
+                modules = self._get_selected_modules(usage_info.user, usage_info.course_key, block_key)
+                for module in modules:
+                    module_state = module.state
+                    state_dict = json.loads(module_state)
+                    # Check all selected entries for this user on
+                    # selected library.
+                    # Add all selected to selected list.
+                    for state in state_dict['selected']:
+                        usage_key = usage_info.course_key.make_usage_key(state[0], state[1])
+                        if usage_key in library_children:
+                            selected.append((state[0], state[1]))
+
+                # update selected
+                previous_count = len(selected)
+                block_keys = LibraryContentModule.make_selection(selected, library_children, max_count, mode)
+                selected = block_keys['selected']
+
+                # publish events for analytics
+                self._publish_events(block_structure, block_key, previous_count, max_count, block_keys)
+                all_selected_children.update(usage_info.course_key.make_usage_key(s[0], s[1]) for s in selected)
+
+        def check_child_removal(block_key):
+            """
+            Check if selected block should be removed.
+
+            Block is removed if it is part of library_content, but has
+            not been selected for current user.
+            """
+            if block_key not in all_library_children:
+                return False
+            if block_key in all_selected_children:
+                return False
+            return True
+
+        # Check and remove all non-selected children from course
+        # structure.
+        block_structure.remove_block_if(
+            check_child_removal
+        )
+
+    @classmethod
+    def _get_selected_modules(cls, user, course_key, block_key):
+        """
+        Get list of selected modules in a library for the given user.
+
+        Arguments:
+            user (User)
+            course_key (CourseLocator)
+            block_key (BlockUsageLocator)
+
+        Returns:
+            list[modules]
+        """
+        return StudentModule.objects.filter(
+            student=user,
+            course_id=course_key,
+            module_state_key=block_key,
+            state__contains='"selected": [['
```

Comment:
It's been a while since I worked on this code, but:
* The docstring of this method is misleading. All this does is return the StudentModule of `block_key` for `user`. Getting the actual list of selected module IDs must be done by parsing the JSON in the `.state` property of the returned object, and then reading IDs out of the `"selected"` value in that JSON, which is a list-of-lists. The code for doing that should be moved into this method - right now it is elsewhere, where this method is called.
* `StudentModule` is uniquely keyed on (`student`, `module_state_key`, `course_id`). So this should be an `try: StudentModule.objects.get(); except StudentModule.DoesNotExist` query and not a `filter()` query. This method also should not be returning a list unless you refactor it to match the docstring.
* What is the `state__contains='"selected": [['` filter supposed to achieve? When would a LibraryContent StudentModule exist but not have a `selected` field? I don't think that condition is adding any value here.